### DISCUSSION TITLE
How to conditionally update?

### DISCUSSION THREAD

**Discussion by greenking19** [simolus3/drift#1941] [Q&A]
I have some data to be fetched from the server, compare it against the local data, and then judge whether to insert or update.

table cloumn name 👇
```dart
TextColumn get updateStatus => text()();
```
get server list data
```dart
var listData = await getListData();
// if tbl.updateStatus == '1'  do update
// if tbl.updateStatus == '0' do nothing
// If this data table does not exist, insert it
for(var item in listData){
    await db.select(db.testTable)...
    // ....I don't know how to write more conveniently
}
```

**Comment by simolus3** [comment]
In SQL, you can use something called an [upsert](https://www.sqlite.org/lang_UPSERT.html) for this. I recommend reading the article from sqlite's documentation, it gives a good introduction of the SQL behind this and is helpful to understand.
With an upsert, it is possible to make an insert sometimes be an insert, and sometimes an update.

I assume that your `testTable` has some id column with a unique constraint (a primary key will do), right? That's all you need to use upserts. You can then do something like

```dart
await db.into(db.testTable).insert(
  yourFetchedRow,
  onConflict: DoUpdate(
    (old) => yourFetchedRow,
    where: (old) => old.updateStatus.equalsValue('1')
  ),
);
```

This essentially means that

- you want to insert an instance, `yourFetchedRow`
- if there's an conflict (because a row with the same primary key already exists), you instead do an update
 - the `(old) => yourFetchedRow` means: Update all columns in the row with entries from the row you want to insert, essentially replacing the row
 - but, the `where` filter means: Only do this if the `updateStatus` of the existing row has a value of `1`. Otherwise, make the whole operation a no-op.
